sys/net/gnrc/network_layer/ipv6/nib/nib.c

```diff
@@ -178,15 +179,25 @@ static bool _on_link(const ipv6_addr_t *dst, unsigned *iface)
     }
 #endif  /* CONFIG_GNRC_IPV6_NIB_6LN */
     while ((entry = _nib_offl_iter(entry))) {
-        if ((entry->mode & _PL) && (entry->flags & _PFX_ON_LINK) &&
-            (ipv6_addr_match_prefix(dst, &entry->pfx) >= entry->pfx_len) &&
+        if ((ipv6_addr_match_prefix(dst, &entry->pfx) >= entry->pfx_len) &&
             (entry->pfx_len > best_pfx)) {
+
+            /* prefix is on-link */
+            if ((entry->mode & _PL) && (entry->flags & _PFX_ON_LINK)) {
+                on_link = true;
+            /* prefix is behind another hop */
+            } else if (entry->mode & _FT) {
+                on_link = false;
+            } else {
+                continue;
+            }
+
             *iface = _nib_onl_get_if(entry->next_hop);
```

Comment:
BTW there might be an (unrelated) bug here. As for 6Lo-ND only link-local addresses are on-link, `best_pfx` should only be set if `*iface` is not a 6LN interface (this case is checked above, but only when `*iface` was initialized before the call.

Reply:
Huh, if the WPAN got a prefix from the border router, all addresses with that prefix should be on-link too, no?

Reply:
Ah I just remembered: No the prefix should not be advertised as on-link in the first place: https://datatracker.ietf.org/doc/html/rfc6775#section-5.4

Reply:
(see also https://datatracker.ietf.org/doc/html/rfc6775#section-6.1)